File: src/8장/8-4.py
from collections import deque

def direction_change(d, c):
    if c == "L":
        d = (d - 1) % 4
    else:
        d = (d + 1) % 4
    return d

N = int(input())
K = int(input())
board = [[0] * N for _ in range(N)]
for _ in range(K):
    a, b = map(int, input().split())
    board[a - 1][b - 1] = 1  
L = int(input())
times = {}
for i in range(L):
    X, C = input().split()
    times[int(X)] = C

dy = [-1, 0, 1, 0]
dx = [0, 1, 0, -1]

direction = 1 
time = 1  
y, x = 0, 0  
snake = deque([[y, x]]) 
board[y][x] = 2

while True:
    y, x = y + dy[direction], x + dx[direction]
    if 0 <= y < N and 0 <= x < N and board[y][x] != 2:
        if not board[y][x] == 1:
            del_y, del_x = snake.popleft()
            board[del_y][del_x] = 0 
        board[y][x] = 2
        snake.append([y, x])
        if time in times.keys():
            direction = direction_change(direction, times[time])
        time += 1
    else: 
        break

print(time)

File: src/8장/8-4.java
import java.util.*;
import java.io.*;

public class Main {

	static Queue<Point> times = new LinkedList<>();
	static Deque<Point> snakeDeque = new ArrayDeque<>();
	static int direction = 3;
	static int[][] board;
	static int[] dx = { -1, 0, 1, 0 };
	static int[] dy = { 0, -1, 0, 1 };

	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st;

		int N = Integer.parseInt(br.readLine());
		int K = Integer.parseInt(br.readLine());

		board = new int[N][N];
		board[0][0] = -1;
		snakeDeque.offer(new Point(0, 0));

		for (int i = 0; i < K; i++) {
			st = new StringTokenizer(br.readLine());
			int a = Integer.parseInt(st.nextToken()) - 1;
			int b = Integer.parseInt(st.nextToken()) - 1;
			board[a][b] = 1;
		}
		int L = Integer.parseInt(br.readLine());
		for (int i = 0; i < L; i++) {
			st = new StringTokenizer(br.readLine());
			int x = Integer.parseInt(st.nextToken());
			String s = st.nextToken();
			int c;
			if (s.equals("L"))
				c = 0;
			else
				c = 1;
			times.add(new Point(x, c));
		}

		int time = 0;
		int X = times.peek().x;
		int C = times.poll().y;

		while (true) {
			time++;
			int nx = snakeDeque.peekFirst().x + dx[direction];
			int ny = snakeDeque.peekFirst().y + dy[direction];
			if (nx < 0 || nx >= N || ny < 0 || ny >= N || board[nx][ny] == -1) {
				break;
			}
			if (board[nx][ny] == 1) {
				board[nx][ny] = -1;
				snakeDeque.offerFirst(new Point(nx, ny));
			} else {
				board[nx][ny] = -1;
				snakeDeque.offerFirst(new Point(nx, ny));
				board[snakeDeque.peekLast().x][snakeDeque.pollLast().y] = 0;
			}
			if (time == X) {
				if (C == 0) {
					direction = (direction + 1) % 4;
				} else {
					direction = (direction + 3) % 4;
				}
				if (!times.isEmpty()) {
					X = times.peek().x;
					C = times.poll().y;
				}
			}
		}
		System.out.println(time);
		br.close();
	}

}

class Point {
	int x;
	int y;

	Point(int x, int y) {
		this.x = x;
		this.y = y;
	}
}